File: tests/conformance/ConformanceTests.cs
// pdftract SDK Conformance Test Runner (.NET / C#)
//
// This test runs the shared SDK conformance suite against the .NET SDK.
// It loads tests/sdk-conformance/cases.json and executes each test case.
//
// Run with: dotnet test --filter ConformanceTests
// Or as standalone: dotnet run --project ConformanceTests.csproj

using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Nodes;
using Xunit;
using Xunit.Abstractions;

namespace Pdftract.Tests
{
    public class ConformanceTests
    {
        private const string SuitePath = "tests/sdk-conformance/cases.json";
        private const string SdkName = "pdftract-dotnet";
        private const string SdkVersion = "0.1.0";

        private readonly ITestOutputHelper _output;

        public ConformanceTests(ITestOutputHelper output)
        {
            _output = output;
        }

        private enum TestStatus
        {
            Pass,
            Fail,
            Skip,
            Error
        }

        private class TestResult
        {
            public string Id { get; set; } = string.Empty;
            public TestStatus Status { get; set; }
            public JsonNode? Actual { get; set; }
            public JsonNode? Expected { get; set; }
            public string? Error { get; set; }
            public string? Reason { get; set; }
            public long DurationMs { get; set; }
        }

        private class ConformanceReport
        {
            public string Sdk { get; set; } = SdkName;
            public string SdkVersion { get; set; } = SdkVersion;
            public string SuiteVersion { get; set; } = string.Empty;
            public string SchemaVersion { get; set; } = string.Empty;
            public string Timestamp { get; set; } = DateTime.UtcNow.ToString("o");
            public List<TestResult> Results { get; set; } = new();
            public Summary Summary { get; set; } = new();
            public Environment Environment { get; set; } = new();
        }

        private class Summary
        {
            public int Total { get; set; }
            public int Passed { get; set; }
            public int Failed { get; set; }
            public int Skipped { get; set; }
            public int Errors { get; set; }
            public long DurationMs { get; set; }
        }

        private class Environment
        {
            public string Os { get; set; } = Environment.OSVersion.Platform.ToString();
            public string Arch { get; set; } = Environment.Is64BitProcess ? "x64" : "x86";
            public string BinaryVersion { get; set; } = SdkVersion;
            public string RuntimeVersion { get; set; } = Environment.Version.ToString();
        }

        private bool CompareWithTolerance(double actual, double expected, JsonObject? tolerance)
        {
            if (tolerance == null)
            {
                return Math.Abs(actual - expected) < 1e-9;
            }

            if (tolerance.TryGetValue("abs", out JsonNode? absNode) && absNode != null)
            {
                double absTol = absNode.GetValue<double>();
                if (Math.Abs(actual - expected) <= absTol)
                {
                    return true;
                }
            }

            if (tolerance.TryGetValue("rel", out JsonNode? relNode) && relNode != null)
            {
                double relTol = relNode.GetValue<double>();
                double diff = Math.Abs(actual - expected);
                double avg = (actual + expected) / 2.0;
                if (avg > 0.0 && diff / avg <= relTol)
                {
                    return true;
                }
            }

            return false;
        }

        private JsonObject? FindTolerance(JsonObject? tolerances, string path)
        {
            if (tolerances == null) return null;

            if (tolerances.TryGetValue(path, out JsonNode? value) && value != null)
            {
                return value.AsObject();
            }

            foreach (var kvp in tolerances)
            {
                if (kvp.Key.Contains('*'))
                {
                    var pattern = kvp.Key.Replace("*", ".*");
                    if (System.Text.RegularExpressions.Regex.IsMatch(path, pattern))
                    {
                        return kvp.Value.AsObject();
                    }
                }
            }

            return null;
        }

        private (bool Passed, string? Reason) CompareResults(
            JsonNode actual, JsonNode expected, JsonObject? tolerances, string path = "")
        {
            if (expected is JsonObject expObj)
            {
                if (actual is JsonValue actVal && actVal.TryGetValue(out double? actNum) && actNum != null)
                {
                    if (expObj.TryGetValue("min", out JsonNode? minNode) && minNode != null)
                    {
                        double min = minNode.GetValue<double>();
                        if (actNum.Value < min)
                        {
                            return (false, $"{path}: value {actNum} < minimum {min}");
                        }
                    }
                    if (expObj.TryGetValue("max", out JsonNode? maxNode) && maxNode != null)
                    {
                        double max = maxNode.GetValue<double>();
                        if (actNum.Value > max)
                        {
                            return (false, $"{path}: value {actNum} > maximum {max}");
                        }
                    }
                    if (expObj.TryGetValue("value", out JsonNode? valNode) && valNode != null)
                    {
                        double expVal = valNode.GetValue<double>();
                        var tol = FindTolerance(tolerances, path);
                        if (!CompareWithTolerance(actNum.Value, expVal, tol))
                        {
                            return (false, $"{path}: numeric mismatch");
                        }
                    }
                }
                else if (actual is JsonValue actStrVal && actStrVal.TryGetValue(out string? actStr) && actStr != null)
                {
                    if (expObj.TryGetValue("min_length", out JsonNode? minLenNode) && minLenNode != null)
                    {
                        int minLen = minLenNode.GetValue<int>();
                        if (actStr.Length < minLen)
                        {
                            return (false, $"{path}: string length {actStr.Length} < minimum {minLen}");
                        }
                    }
                    if (expObj.TryGetValue("contains", out JsonNode? containsNode) && containsNode != null)
                    {
                        var contains = containsNode.AsArray();
                        foreach (var item in contains)
                        {
                            if (item.TryGetValue(out string? substr) && substr != null && !actStr.Contains(substr))
                            {
                                return (false, $"{path}: string does not contain '{substr}'");
                            }
                        }
                    }
                }
                else if (actual is JsonArray actArr)
                {
                    if (expObj.TryGetValue("min", out JsonNode? minNode) && minNode != null)
                    {
                        int min = minNode.GetValue<int>();
                        if (actArr.Count < min)
                        {
                            return (false, $"{path}: array length {actArr.Count} < minimum {min}");
                        }
                    }
                    if (expObj.TryGetValue("max", out JsonNode? maxNode) && maxNode != null)
                    {
                        int max = maxNode.GetValue<int>();
                        if (actArr.Count > max)
                        {
                            return (false, $"{path}: array length {actArr.Count} > maximum {max}");
                        }
                    }
                }
                else if (actual is JsonObject actObj)
                {
                    foreach (var kvp in expObj)
                    {
                        var newPath = string.IsNullOrEmpty(path) ? kvp.Key : $"{path}.{kvp.Key}";
                        if (!actObj.TryGetValue(kvp.Key, out JsonNode? actValue))
                        {
                            return (false, $"{newPath}: missing key '{kvp.Key}'");
                        }
                        var (passed, reason) = CompareResults(actValue, kvp.Value!, tolerances, newPath);
                        if (!passed) return (false, reason);
                    }
                }
            }
            else if (expected is JsonArray expArr && actual is JsonArray actArr2)
            {
                for (int i = 0; i < expArr.Count; i++)
                {
                    var newPath = $"{path}[{i}]";
                    if (i >= actArr2.Count)
                    {
                        return (false, $"{newPath}: missing index");
                    }
                    var (passed, reason) = CompareResults(actArr2[i], expArr[i], tolerances, newPath);
                    if (!passed) return (false, reason);
                }
            }
            else
            {
                if (!JsonNode.DeepEquals(actual, expected))
                {
                    return (false, $"{path}: expected {expected.ToJsonString()}, got {actual.ToJsonString()}");
                }
            }

            return (true, null);
        }

        private JsonNode ExecuteMethod(string method, string fixture, JsonObject options)
        {
            // This is a stub - replace with actual SDK calls when available
            return method switch
            {
                "extract" => new JsonObject
                {
                    ["schema_version"] = "1.0",
                    ["metadata"] = new JsonObject { ["page_count"] = 1 },
                    ["pages"] = new JsonArray
                    {
                        new JsonObject
                        {
                            ["page_index"] = 0,
                            ["width"] = 612,
                            ["height"] = 792,
                            ["rotation"] = 0
                        }
                    },
                    ["errors"] = new JsonArray()
                },
                "extract_text" => new JsonValue("Sample text content"),
                "extract_markdown" => new JsonValue("# Sample Markdown\n\nContent here"),
                "hash" => new JsonObject { ["hash"] = "abc123", ["fast_hash"] = "def456" },
                _ => JsonValue.Create(null)
            };
        }

        private int CompareVersions(string v1, string v2)
        {
            var parts1 = v1.Split('.');
            var parts2 = v2.Split('.');

            for (int i = 0; i < Math.Min(parts1.Length, parts2.Length); i++)
            {
                if (int.TryParse(parts1[i], out int n1) && int.TryParse(parts2[i], out int n2))
                {
                    if (n1 < n2) return -1;
                    if (n1 > n2) return 1;
                }
            }

            return parts1.Length.CompareTo(parts2.Length);
        }

        private TestResult RunTestCase(JsonObject testCase, string schemaVersion, string fixturesBase)
        {
            var stopwatch = Stopwatch.StartNew();
            string id = testCase["id"].GetValue<string>();

            // Check min_schema_version
            if (testCase.TryGetValue("min_schema_version", out JsonNode? minVerNode) && minVerNode != null)
            {
                string minVer = minVerNode.GetValue<string>();
                if (CompareVersions(schemaVersion, minVer) < 0)
                {
                    stopwatch.Stop();
                    return new TestResult
                    {
                        Id = id,
                        Status = TestStatus.Skip,
                        Reason = $"Schema version {schemaVersion} < minimum required {minVer}",
                        DurationMs = stopwatch.ElapsedMilliseconds
                    };
                }
            }

            string fixture = testCase["fixture"].GetValue<string>();
            string method = testCase["method"].GetValue<string>();
            var options = testCase["options"].AsObject();
            var expected = testCase["expected"];
            var tolerances = testCase.TryGetValue("tolerances", out JsonNode? tol) ? tol.AsObject() : null;

            string fixturePath = fixture.StartsWith("http") ? fixture :
                Path.Combine(fixturesBase, fixture);

            try
            {
                var actual = ExecuteMethod(method, fixturePath, options);
                var (passed, reason) = CompareResults(actual, expected, tolerances);

                stopwatch.Stop();
                return new TestResult
                {
                    Id = id,
                    Status = passed ? TestStatus.Pass : TestStatus.Fail,
                    Actual = actual,
                    Expected = expected,
                    Reason = reason,
                    DurationMs = stopwatch.ElapsedMilliseconds
                };
            }
            catch (Exception ex)
            {
                stopwatch.Stop();
                return new TestResult
                {
                    Id = id,
                    Status = TestStatus.Error,
                    Expected = expected,
                    Error = ex.Message,
                    DurationMs = stopwatch.ElapsedMilliseconds
                };
            }
        }

        private ConformanceReport RunConformance(string suitePath, string outputPath)
        {
            _output.WriteLine($"pdftract SDK Conformance Runner");
            _output.WriteLine($"SDK: {SdkName} v{SdkVersion}");
            _output.WriteLine($"Suite: {suitePath}");
            _output.WriteLine("");

            var suiteJson = File.ReadAllText(suitePath);
            var suite = JsonNode.Parse(suiteJson)?.AsObject()
                ?? throw new InvalidOperationException("Failed to parse suite");

            string suiteVersion = suite["version"].GetValue<string>();
            string schemaVersion = suite["schema_version"].GetValue<string>();
            var cases = suite["cases"].AsArray();

            string fixturesBase = Path.Combine(Path.GetDirectoryName(suitePath) ?? "", "fixtures");

            _output.WriteLine($"Found {cases.Count} test cases");
            _output.WriteLine("");

            var stopwatch = Stopwatch.StartNew();
            var results = new List<TestResult>();

            foreach (var testCase in cases)
            {
                var result = RunTestCase(testCase!.AsObject(), schemaVersion, fixturesBase);

                _output.WriteLine($"[{result.Status}] {result.Id} ({result.DurationMs}ms)");

                if (result.Status == TestStatus.Fail || result.Status == TestStatus.Error)
                {
                    if (result.Reason != null) _output.WriteLine($"  Reason: {result.Reason}");
                    if (result.Error != null) _output.WriteLine($"  Error: {result.Error}");
                }

                results.Add(result);
            }

            stopwatch.Stop();

            var summary = new Summary
            {
                Total = results.Count,
                Passed = results.Count(r => r.Status == TestStatus.Pass),
                Failed = results.Count(r => r.Status == TestStatus.Fail),
                Skipped = results.Count(r => r.Status == TestStatus.Skip),
                Errors = results.Count(r => r.Status == TestStatus.Error),
                DurationMs = stopwatch.ElapsedMilliseconds
            };

            _output.WriteLine("");
            _output.WriteLine("Summary:");
            _output.WriteLine($"  Total:   {summary.Total}");
            _output.WriteLine($"  Passed:  {summary.Passed}");
            _output.WriteLine($"  Failed:  {summary.Failed}");
            _output.WriteLine($"  Skipped: {summary.Skipped}");
            _output.WriteLine($"  Errors:  {summary.Errors}");
            _output.WriteLine($"  Time:    {summary.DurationMs}ms");

            var report = new ConformanceReport
            {
                SuiteVersion = suiteVersion,
                SchemaVersion = schemaVersion,
                Timestamp = DateTime.UtcNow.ToString("o"),
                Results = results,
                Summary = summary,
                Environment = new Environment()
            };

            File.WriteAllText(outputPath, JsonSerializer.Serialize(report, new JsonSerializerOptions
            {
                WriteIndented = true
            }));

            _output.WriteLine("");
            _output.WriteLine($"Report written to: {outputPath}");

            return report;
        }

        [Fact]
        public void TestConformanceSuite()
        {
            var report = RunConformance(SuitePath, "conformance-report.json");
            Assert.Equal(0, report.Summary.Failed);
            Assert.Equal(0, report.Summary.Errors);
        }
    }
}

File: tests/conformance/ConformanceTests.swift
/*
 * pdftract SDK Conformance Test Runner (Swift)
 *
 * This test runs the shared SDK conformance suite against the Swift SDK.
 * It loads tests/sdk-conformance/cases.json and executes each test case.
 *
 * Run with: swift test --filter ConformanceTests
 * Or as standalone: swift ConformanceTests.swift <suite-path> <output-path>
 */

import Foundation

#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

let SUITE_PATH = "tests/sdk-conformance/cases.json"
let SDK_NAME = "pdftract-swift"
let SDK_VERSION = "0.1.0"

enum TestStatus: String, Encodable {
    case pass = "pass"
    case fail = "fail"
    case skip = "skip"
    case error = "error"
}

struct TestResult: Encodable {
    let id: String
    let status: TestStatus
    let actual: String?
    let expected: String?
    let error: String?
    let reason: String?
    let duration_ms: Int64

    func toDict() -> [String: Any] {
        var dict: [String: Any] = [
            "id": id,
            "status": status.rawValue,
            "duration_ms": duration_ms
        ]
        if let actual = actual { dict["actual"] = actual }
        if let expected = expected { dict["expected"] = expected }
        if let error = error { dict["error"] = error }
        if let reason = reason { dict["reason"] = reason }
        return dict
    }
}

struct Summary: Encodable {
    let total: Int
    let passed: Int
    let failed: Int
    let skipped: Int
    let errors: Int
    let duration_ms: Int64
}

struct Environment: Encodable {
    let os: String
    let arch: String
    let binary_version: String
    let runtime_version: String
}

struct ConformanceReport: Encodable {
    let sdk: String
    let sdk_version: String
    let suite_version: String
    let schema_version: String
    let timestamp: String
    let results: [TestResult]
    let summary: Summary
    let environment: Environment
}

func compareWithTolerance(_ actual: Double, _ expected: Double, _ tolerance: [String: Any]?) -> Bool {
    guard let tolerance = tolerance else {
        return abs(actual - expected) < Double.ulpOfOne
    }

    if let absTol = tolerance["abs"] as? Double {
        if abs(actual - expected) <= absTol {
            return true
        }
    }

    if let relTol = tolerance["rel"] as? Double {
        let diff = abs(actual - expected)
        let avg = (actual + expected) / 2.0
        if avg > 0.0 && diff / avg <= relTol {
            return true
        }
    }

    return false
}

func findTolerance(_ tolerances: [String: Any]?, _ path: String) -> [String: Any]? {
    guard let tolerances = tolerances else { return nil }

    if let val = tolerances[path] {
        return val as? [String: Any]
    }

    for (key, val) in tolerances {
        if key.contains("*") {
            let pattern = key.replacingOccurrences(of: "*", with: ".*")
            if let regex = try? NSRegularExpression(pattern: pattern),
               regex.firstMatch(in: path, range: NSRange(location: 0, length: path.utf16.count)) != nil {
                return val as? [String: Any]
            }
        }
    }

    return nil
}

func compareResults(_ actual: Any, _ expected: Any, _ tolerances: [String: Any]?, _ path: String = "") -> (Bool, String?) {
    if let expDict = expected as? [String: Any] {
        if let actNum = actual as? Double {
            if let min = expDict["min"] as? Double {
                if actNum < min {
                    return (false, "\(path): value \(actNum) < minimum \(min)")
                }
            }
            if let max = expDict["max"] as? Double {
                if actNum > max {
                    return (false, "\(path): value \(actNum) > maximum \(max)")
                }
            }
            if let val = expDict["value"] as? Double {
                let tol = findTolerance(tolerances, path)
                if !compareWithTolerance(actNum, val, tol) {
                    return (false, "\(path): numeric mismatch")
                }
            }
        } else if let actStr = actual as? String {
            if let minLen = expDict["min_length"] as? Int {
                if actStr.count < minLen {
                    return (false, "\(path): string length too short")
                }
            }
            if let contains = expDict["contains"] as? [String] {
                for substring in contains {
                    if !actStr.contains(substring) {
                        return (false, "\(path): string does not contain '\(substring)'")
                    }
                }
            }
        } else if let actArray = actual as? [Any] {
            if let min = expDict["min"] as? Int {
                if actArray.count < min {
                    return (false, "\(path): array length too short")
                }
            }
            if let max = expDict["max"] as? Int {
                if actArray.count > max {
                    return (false, "\(path): array length too long")
                }
            }
        } else if let actDict = actual as? [String: Any] {
            for (key, expVal) in expDict {
                let newPath = path.isEmpty ? key : "\(path).\(key)"
                guard let actVal = actDict[key] else {
                    return (false, "\(newPath): missing key '\(key)'")
                }
                let (passed, reason) = compareResults(actVal, expVal, tolerances, newPath)
                if !passed {
                    return (false, reason)
                }
            }
        }
    } else if let expArray = expected as? [Any], let actArray = actual as? [Any] {
        for (i, expVal) in expArray.enumerated() {
            let newPath = "\(path)[\(i)]"
            if i >= actArray.count {
                return (false, "\(newPath): missing index")
            }
            let (passed, reason) = compareResults(actArray[i], expVal, tolerances, newPath)
            if !passed {
                return (false, reason)
            }
        }
    } else {
        // Simple comparison
        if let actualStr = actual as? String,
           let expectedStr = expected as? String,
           actualStr != expectedStr {
            return (false, "\(path): strings do not match")
        }
    }

    return (true, nil)
}

func executeMethod(_ method: String, _ fixture: String, _ options: [String: Any]) -> Any {
    // This is a stub - replace with actual SDK calls when available
    switch method {
    case "extract":
        return [
            "schema_version": "1.0",
            "metadata": ["page_count": 1],
            "pages": [
                [
                    "page_index": 0,
                    "width": 612,
                    "height": 792,
                    "rotation": 0
                ]
            ],
            "errors": []
        ] as [String: Any]
    case "extract_text":
        return "Sample text content"
    case "extract_markdown":
        return "# Sample Markdown\n\nContent here"
    case "hash":
        return ["hash": "abc123", "fast_hash": "def456"]
    default:
        return [:] as [String: Any]
    }
}

func compareVersions(_ v1: String, _ v2: String) -> ComparisonResult {
    let parts1 = v1.split(separator: ".").compactMap { Int($0) }
    let parts2 = v2.split(separator: ".").compactMap { Int($0) }

    let maxCount = max(parts1.count, parts2.count)

    for i in 0..<maxCount {
        let n1 = i < parts1.count ? parts1[i] : 0
        let n2 = i < parts2.count ? parts2[i] : 0

        if n1 < n2 {
            return .orderedAscending
        }
        if n1 > n2 {
            return .orderedDescending
        }
    }

    return .orderedSame
}

func runTestCase(_ case: [String: Any], _ schemaVersion: String, _ fixturesBase: String) -> TestResult {
    let start = Date()

    guard let id = case["id"] as? String else {
        return TestResult(
            id: "unknown",
            status: .error,
            actual: nil,
            expected: nil,
            error: "Missing test case ID",
            reason: nil,
            duration_ms: 0
        )
    }

    // Check min_schema_version
    if let minVer = case["min_schema_version"] as? String {
        if compareVersions(schemaVersion, minVer) == .orderedAscending {
            return TestResult(
                id: id,
                status: .skip,
                actual: nil,
                expected: nil,
                error: nil,
                reason: "Schema version \(schemaVersion) < minimum required \(minVer)",
                duration_ms: Int64(Date().timeIntervalSince(start) * 1000)
            )
        }
    }

    guard let fixture = case["fixture"] as? String,
          let method = case["method"] as? String else {
        return TestResult(
            id: id,
            status: .error,
            actual: nil,
            expected: nil,
            error: "Missing required fields",
            reason: nil,
            duration_ms: 0
        )
    }

    let options = case["options"] as? [String: Any] ?? [:]
    let expected = case["expected"] ?? [:]
    let tolerances = case["tolerances"] as? [String: Any]

    let fixturePath: String
    if fixture.hasPrefix("http") {
        fixturePath = fixture
    } else {
        fixturePath = "\(fixturesBase)/\(fixture)"
    }

    do {
        let actual = executeMethod(method, fixturePath, options)
        let (passed, reason) = compareResults(actual, expected, tolerances)

        if passed {
            return TestResult(
                id: id,
                status: .pass,
                actual: String(describing: actual),
                expected: String(describing: expected),
                error: nil,
                reason: nil,
                duration_ms: Int64(Date().timeIntervalSince(start) * 1000)
            )
        } else {
            return TestResult(
                id: id,
                status: .fail,
                actual: String(describing: actual),
                expected: String(describing: expected),
                error: nil,
                reason: reason,
                duration_ms: Int64(Date().timeIntervalSince(start) * 1000)
            )
        }
    } catch {
        return TestResult(
            id: id,
            status: .error,
            actual: nil,
            expected: String(describing: expected),
            error: String(describing: error),
            reason: nil,
            duration_ms: Int64(Date().timeIntervalSince(start) * 1000)
        )
    }
}

func runConformance(_ suitePath: String, _ outputPath: String) -> ConformanceReport {
    print("pdftract SDK Conformance Runner")
    print("SDK: \(SDK_NAME) v\(SDK_VERSION)")
    print("Suite: \(suitePath)")
    print("")

    guard let suiteData = try? Data(contentsOf: URL(fileURLWithPath: suitePath)),
          let suite = try? JSONSerialization.jsonObject(with: suiteData) as? [String: Any] else {
        fatalError("Failed to load suite")
    }

    let suiteVersion = suite["version"] as? String ?? "unknown"
    let schemaVersion = suite["schema_version"] as? String ?? "unknown"
    let cases = suite["cases"] as? [[String: Any]] ?? []

    let fixturesBase = ((suitePath as NSString).deletingLastPathComponent as NSString).appendingPathComponent("fixtures")

    print("Found \(cases.count) test cases")
    print("")

    let start = Date()
    var results: [TestResult] = []

    for testCase in cases {
        let result = runTestCase(testCase, schemaVersion, fixturesBase)
        results.append(result)

        let statusSym: String
        switch result.status {
        case .pass: statusSym = "PASS"
        case .fail: statusSym = "FAIL"
        case .skip: statusSym = "SKIP"
        case .error: statusSym = "ERROR"
        }

        print("[\(statusSym)] \(result.id) (\(result.duration_ms)ms)")

        if result.status == .fail || result.status == .error {
            if let reason = result.reason {
                print("  Reason: \(reason)")
            }
            if let error = result.error {
                print("  Error: \(error)")
            }
        }
    }

    let duration_ms = Int64(Date().timeIntervalSince(start) * 1000)

    let passed = results.filter { $0.status == .pass }.count
    let failed = results.filter { $0.status == .fail }.count
    let skipped = results.filter { $0.status == .skip }.count
    let errors = results.filter { $0.status == .error }.count

    print("")
    print("Summary:")
    print("  Total:   \(results.count)")
    print("  Passed:  \(passed)")
    print("  Failed:  \(failed)")
    print("  Skipped: \(skipped)")
    print("  Errors:  \(errors)")
    print("  Time:    \(duration_ms)ms")

    let report = ConformanceReport(
        sdk: SDK_NAME,
        sdk_version: SDK_VERSION,
        suite_version: suiteVersion,
        schema_version: schemaVersion,
        timestamp: ISO8601DateFormatter().string(from: Date()),
        results: results,
        summary: Summary(
            total: results.count,
            passed: passed,
            failed: failed,
            skipped: skipped,
            errors: errors,
            duration_ms: duration_ms
        ),
        environment: Environment(
            os: "macOS", // Runtime detection would go here
            arch: "arm64",
            binary_version: SDK_VERSION,
            runtime_version: "5.9"
        )
    )

    if let reportData = try? JSONEncoder().encode(report),
       let reportJson = String(data: reportData, encoding: .utf8) {
        try? reportJson.write(toFile: outputPath, atomically: true, encoding: .utf8)
        print("")
        print("Report written to: \(outputPath)")
    }

    return report
}

// CLI entry point
if CommandLine.argc > 1 {
    let suiteArg = CommandLine.arguments.count > 1 ? CommandLine.arguments[1] : SUITE_PATH
    let outputArg = CommandLine.arguments.count > 2 ? CommandLine.arguments[2] : "conformance-report.json"

    let report = runConformance(suiteArg, outputArg)

    exit(report.summary.failed == 0 && report.summary.errors == 0 ? 0 : 1)
}
